File: backend/app/routers/keys.py
import re
from datetime import datetime, timezone

import httpx
from fastapi import APIRouter, Depends, Form, HTTPException, Request
from fastapi.responses import HTMLResponse, RedirectResponse
from sqlalchemy import desc, func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app import cache
from app.config import settings
from app.db import get_db
from app.deps import current_user, current_user_optional
from app.models import Alert, ApiKey, KeyLimit, KeyUsage, User
from app.providers import PROVIDERS, auth_headers, probe_path, resolve_base_url
from app.security import decrypt_secret, encrypt_secret, hash_secret
from app.templating import templates
from app.usage import cached_spend

router = APIRouter(tags=["keys"])

ALIAS_RE = re.compile(r"^[a-zA-Z0-9_-]{2,32}$")


async def _owned_key(db: AsyncSession, user: User, alias: str) -> ApiKey:
    result = await db.execute(
        select(ApiKey).where(ApiKey.user_id == user.id, ApiKey.alias == alias)
    )
    key = result.scalar_one_or_none()
    if key is None:
        raise HTTPException(status_code=404, detail="Alias not found")
    return key


async def _next_alias(db: AsyncSession, user_id: int) -> str:
    rows = (
        await db.execute(select(ApiKey.alias).where(ApiKey.user_id == user_id))
    ).scalars().all()
    used = {int(m.group(1)) for a in rows if (m := re.fullmatch(r"key(\d+)", a))}
    n = 1
    while n in used:
        n += 1
    return f"key{n}"


async def _key_view(db: AsyncSession, key: ApiKey) -> dict:
    limit = (
        await db.execute(select(KeyLimit).where(KeyLimit.api_key_id == key.id))
    ).scalar_one_or_none()
    spend = await cached_spend(db, key.id)
    requests_today = await db.scalar(
        select(func.count(KeyUsage.id)).where(
            KeyUsage.api_key_id == key.id,
            KeyUsage.created_at
            >= datetime.now(timezone.utc).replace(hour=0, minute=0, second=0, microsecond=0),
        )
    )
    rate_used = (
        await cache.peek_rate(key.id, limit.rate_window_seconds) if limit and limit.rate_limit else 0
    )
    return {
        "key": key,
        "limit": limit,
        "spend": spend,
        "spend_pct": min(100, int(spend / limit.spend_cap_usd * 100)) if limit and limit.spend_cap_usd else 0,
        "rate_used": rate_used,
        "rate_pct": min(100, int(rate_used / limit.rate_limit * 100)) if limit and limit.rate_limit else 0,
        "requests_today": requests_today or 0,
    }


# ---------------- dashboard ----------------

@router.get("/", response_class=HTMLResponse)
async def dashboard(
    request: Request,
    user: User | None = Depends(current_user_optional),
    db: AsyncSession = Depends(get_db),
):
    if user is None:
        return RedirectResponse("/login", status_code=303)

    keys = (
        await db.execute(
            select(ApiKey).where(ApiKey.user_id == user.id).order_by(ApiKey.id.asc())
        )
    ).scalars().all()

    active = [await _key_view(db, k) for k in keys if k.active]
    revoked = [await _key_view(db, k) for k in keys if not k.active]
    unread = await db.scalar(
        select(func.count(Alert.id)).where(Alert.user_id == user.id, Alert.read.is_(False))
    )

    return templates.TemplateResponse(
        request,
        "dashboard.html",
        {"user": user, "active_keys": active, "revoked_keys": revoked, "unread": unread or 0},
    )


# ---------------- create ----------------

@router.get("/keys/new", response_class=HTMLResponse)
async def new_key_page(
    request: Request, user: User = Depends(current_user), db: AsyncSession = Depends(get_db)
):
    return templates.TemplateResponse(
        request,
        "key_new.html",
        {
            "user": user,
            "providers": PROVIDERS,
            "suggested_alias": await _next_alias(db, user.id),
            "created": None,
            "error": None,
        },
    )


@router.post("/keys/new", response_class=HTMLResponse)
async def create_key(
    request: Request,
    secret: str = Form(...),
    provider: str = Form("openai"),
    label: str = Form(""),
    alias: str = Form(""),
    base_url: str = Form(""),
    passphrase: str = Form(""),
    rate_limit: int = Form(0),
    rate_window_seconds: int = Form(60),
    spend_cap_usd: float = Form(0.0),
    mode: str = Form("block"),
    user: User = Depends(current_user),
    db: AsyncSession = Depends(get_db),
):
    alias = (alias or "").strip() or await _next_alias(db, user.id)
    error = None
    if not ALIAS_RE.match(alias):
        error = "Alias must be 2-32 characters: letters, numbers, dashes or underscores."
    elif (
        await db.execute(select(ApiKey).where(ApiKey.user_id == user.id, ApiKey.alias == alias))
    ).scalar_one_or_none():
        error = f"You already have an alias called '{alias}'."
    elif not secret.strip():
        error = "Paste the provider API key."
    elif provider == "custom" and not base_url.strip():
        error = "A custom provider needs a base URL."

    if error:
        return templates.TemplateResponse(
            request,
            "key_new.html",
            {
                "user": user,
                "providers": PROVIDERS,
                "suggested_alias": alias,
                "created": None,
                "error": error,
            },
            status_code=400,
        )

    secret = secret.strip()
    key = ApiKey(
        user_id=user.id,
        alias=alias,
        label=label.strip(),
        provider=provider,
        base_url=base_url.strip(),
        secret_ciphertext=encrypt_secret(secret),
        secret_last4=secret[-4:],
        passphrase_hash=hash_secret(passphrase.strip()) if passphrase.strip() else None,
        active=True,
    )
    db.add(key)
    await db.flush()

    db.add(
        KeyLimit(
            api_key_id=key.id,
            rate_limit=max(0, rate_limit),
            rate_window_seconds=max(1, rate_window_seconds),
            rate_mode=mode,
            spend_cap_usd=max(0.0, spend_cap_usd),
            spend_mode=mode,
        )
    )
    await db.commit()
    await db.refresh(key)

    return templates.TemplateResponse(
        request,
        "key_new.html",
        {
            "user": user,
            "providers": PROVIDERS,
            "suggested_alias": await _next_alias(db, user.id),
            "created": key,
            "has_passphrase": bool(passphrase.strip()),
            "error": None,
        },
    )


# ---------------- test connection ----------------

@router.post("/api/keys/{alias}/test", response_class=HTMLResponse)
async def test_connection(
    request: Request,
    alias: str,
    user: User = Depends(current_user),
    db: AsyncSession = Depends(get_db),
):
    key = await _owned_key(db, user, alias)
    if not key.active:
        return HTMLResponse(
            '<span class="badge badge-bad">Revoked &mdash; re-activate this alias first</span>'
        )

    base = resolve_base_url(key.provider, key.base_url)
    url = f"{base}{probe_path(key.provider)}"
    headers = auth_headers(key.provider, decrypt_secret(key.secret_ciphertext))

    started = datetime.now(timezone.utc)
    try:
        async with httpx.AsyncClient(timeout=15) as client:
            response = await client.get(url, headers=headers)
    except httpx.HTTPError as exc:
        return HTMLResponse(
            f'<span class="badge badge-bad">Could not reach provider &mdash; {exc.__class__.__name__}</span>'
        )

    ms = int((datetime.now(timezone.utc) - started).total_seconds() * 1000)

    db.add(
        KeyUsage(
            api_key_id=key.id,
            path=probe_path(key.provider),
            model="",
            status_code=response.status_code,
            cost_usd=0.0,
            latency_ms=ms,
            is_test=True,
            client_ip="dashboard",
        )
    )
    await db.commit()

    if response.status_code < 400:
        return HTMLResponse(
            f'<span class="badge badge-good">Success &mdash; provider responded '
            f'{response.status_code} in {ms}ms</span>'
        )

    detail = response.text[:180].replace("<", "&lt;")
    return HTMLResponse(
        f'<span class="badge badge-bad">Provider returned {response.status_code}</span>'
        f'<pre class="probe-error">{detail}</pre>'
    )


# ---------------- detail / settings ----------------

@router.get("/keys/{alias}", response_class=HTMLResponse)
async def key_detail(
    request: Request,
    alias: str,
    user: User = Depends(current_user),
    db: AsyncSession = Depends(get_db),
):
    key = await _owned_key(db, user, alias)
    view = await _key_view(db, key)
    recent = (
        await db.execute(
            select(KeyUsage)
            .where(KeyUsage.api_key_id == key.id)
            .order_by(desc(KeyUsage.created_at))
            .limit(50)
        )
    ).scalars().all()
    key_alerts = (
        await db.execute(
            select(Alert)
            .where(Alert.api_key_id == key.id)
            .order_by(desc(Alert.created_at))
            .limit(20)
        )
    ).scalars().all()

    return templates.TemplateResponse(
        request,
        "key_detail.html",
        {"user": user, "view": view, "recent": recent, "alerts": key_alerts, "providers": PROVIDERS},
    )


@router.post("/keys/{alias}/limits")
async def update_limits(
    alias: str,
    rate_limit: int = Form(0),
    rate_window_seconds: int = Form(60),
    rate_mode: str = Form("block"),
    spend_cap_usd: float = Form(0.0),
    spend_mode: str = Form("block"),
    user: User = Depends(current_user),
    db: AsyncSession = Depends(get_db),
):
    key = await _owned_key(db, user, alias)
    limit = (
        await db.execute(select(KeyLimit).where(KeyLimit.api_key_id == key.id))
    ).scalar_one_or_none()
    if limit is None:
        limit = KeyLimit(api_key_id=key.id)
        db.add(limit)

    limit.rate_limit = max(0, rate_limit)
    limit.rate_window_seconds = max(1, rate_window_seconds)
    limit.rate_mode = rate_mode
    limit.spend_cap_usd = max(0.0, spend_cap_usd)
    limit.spend_mode = spend_mode
    await db.commit()
    return RedirectResponse(f"/keys/{alias}", status_code=303)


@router.post("/keys/{alias}/passphrase")
async def update_passphrase(
    alias: str,
    passphrase: str = Form(""),
    user: User = Depends(current_user),
    db: AsyncSession = Depends(get_db),
):
    key = await _owned_key(db, user, alias)
    key.passphrase_hash = hash_secret(passphrase.strip()) if passphrase.strip() else None
    await db.commit()
    return RedirectResponse(f"/keys/{alias}", status_code=303)


@router.post("/keys/{alias}/rename")
async def rename_alias(
    alias: str,
    new_alias: str = Form(...),
    label: str = Form(""),
    user: User = Depends(current_user),
    db: AsyncSession = Depends(get_db),
):
    key = await _owned_key(db, user, alias)
    new_alias = new_alias.strip()
    if ALIAS_RE.match(new_alias):
        clash = (
            await db.execute(
                select(ApiKey).where(
                    ApiKey.user_id == user.id, ApiKey.alias == new_alias, ApiKey.id != key.id
                )
            )
        ).scalar_one_or_none()
        if clash is None:
            key.alias = new_alias
    key.label = label.strip()
    await db.commit()
    return RedirectResponse(f"/keys/{key.alias}", status_code=303)


# ---------------- soft delete ----------------

@router.post("/keys/{alias}/revoke")
async def revoke(
    alias: str, user: User = Depends(current_user), db: AsyncSession = Depends(get_db)
):
    """Soft delete only: flip the flag, keep every usage row and alert."""
    key = await _owned_key(db, user, alias)
    key.active = False
    key.revoked_at = datetime.now(timezone.utc)
    await db.commit()
    return RedirectResponse("/", status_code=303)


@router.post("/keys/{alias}/activate")
async def activate(
    alias: str, user: User = Depends(current_user), db: AsyncSession = Depends(get_db)
):
    key = await _owned_key(db, user, alias)
    key.active = True
    key.revoked_at = None
    await db.commit()
    return RedirectResponse(f"/keys/{alias}", status_code=303)


@router.post("/keys/{alias}/rotate")
async def rotate(
    alias: str,
    secret: str = Form(...),
    user: User = Depends(current_user),
    db: AsyncSession = Depends(get_db),
):
    key = await _owned_key(db, user, alias)
    secret = secret.strip()
    if secret:
        key.secret_ciphertext = encrypt_secret(secret)
        key.secret_last4 = secret[-4:]
        await db.commit()
    return RedirectResponse(f"/keys/{alias}", status_code=303)


@router.get("/keys/{alias}/curl", response_class=HTMLResponse)
async def curl_snippet(
    alias: str, user: User = Depends(current_user), db: AsyncSession = Depends(get_db)
):
    key = await _owned_key(db, user, alias)
    base = settings.public_base_url.rstrip("/")
    token = f"{key.alias}<passphrase>" if key.passphrase_hash else key.alias
    return HTMLResponse(
        f"curl {base}/proxy/{key.alias}/v1/chat/completions "
        f'-H "Authorization: Bearer {token}"'
    )

File: src/routes/keys.js
const express = require("express");
const { Op } = require("sequelize");
const { settings } = require("../config");
const { ApiKey, KeyLimit, KeyUsage, Alert } = require("../models");
const { requireAuth } = require("../middleware/auth");
const { asyncHandler } = require("../middleware/asyncHandler");
const { HttpError } = require("../httpError");
const { PROVIDERS, authHeaders, probePath, resolveBaseUrl } = require("../providers");
const { decryptSecret, encryptSecret, hashSecret } = require("../security");
const { cachedSpend } = require("../usage");
const cache = require("../cache");

const router = express.Router();

const ALIAS_RE = /^[a-zA-Z0-9_-]{2,32}$/;

async function ownedKey(user, alias) {
  const key = await ApiKey.findOne({ where: { userId: user.id, alias } });
  if (!key) throw new HttpError(404, "Alias not found");
  return key;
}

async function nextAlias() {
  // Aliases are globally unique (the proxy route has no user context to
  // scope by), so the suggestion has to avoid every alias in the system,
  // not just this user's.
  const rows = await ApiKey.findAll({ attributes: ["alias"], raw: true });
  const used = new Set();
  for (const { alias } of rows) {
    const m = /^key(\d+)$/.exec(alias);
    if (m) used.add(parseInt(m[1], 10));
  }
  let n = 1;
  while (used.has(n)) n += 1;
  return `key${n}`;
}

async function keyView(key) {
  const limit = await KeyLimit.findOne({ where: { apiKeyId: key.id } });
  const spend = await cachedSpend(key.id);

  const now = new Date();
  const todayStart = new Date(
    Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate(), 0, 0, 0, 0)
  );
  const requestsToday = await KeyUsage.count({
    where: { apiKeyId: key.id, createdAt: { [Op.gte]: todayStart } },
  });

  const rateUsed =
    limit && limit.rateLimit ? await cache.peekRate(key.id, limit.rateWindowSeconds) : 0;

  return {
    key,
    limit,
    spend,
    spend_pct:
      limit && limit.spendCapUsd
        ? Math.min(100, Math.floor((spend / limit.spendCapUsd) * 100))
        : 0,
    rate_used: rateUsed,
    rate_pct:
      limit && limit.rateLimit ? Math.min(100, Math.floor((rateUsed / limit.rateLimit) * 100)) : 0,
    requests_today: requestsToday || 0,
  };
}

// ---------------- dashboard ----------------

router.get(
  "/",
  asyncHandler(async (req, res) => {
    if (!req.user) return res.redirect(303, "/login");

    const keys = await ApiKey.findAll({
      where: { userId: req.user.id },
      order: [["id", "ASC"]],
    });

    const active = [];
    const revoked = [];
    for (const k of keys) {
      const view = await keyView(k);
      (k.active ? active : revoked).push(view);
    }

    const unread = await Alert.count({ where: { userId: req.user.id, read: false } });

    res.render("dashboard.html", {
      user: req.user,
      active_keys: active,
      revoked_keys: revoked,
      unread: unread || 0,
    });
  })
);

// ---------------- create ----------------

router.get(
  "/keys/new",
  requireAuth,
  asyncHandler(async (req, res) => {
    res.render("key_new.html", {
      user: req.user,
      providers: PROVIDERS,
      suggested_alias: await nextAlias(),
      created: null,
      error: null,
    });
  })
);

router.post(
  "/keys/new",
  requireAuth,
  asyncHandler(async (req, res) => {
    const body = req.body;
    const provider = body.provider || "openai";
    const label = body.label || "";
    const baseUrl = body.base_url || "";
    const passphrase = body.passphrase || "";
    const rateLimit = parseInt(body.rate_limit, 10) || 0;
    const rateWindowSeconds = parseInt(body.rate_window_seconds, 10) || 60;
    const spendCapUsd = parseFloat(body.spend_cap_usd) || 0.0;
    const mode = body.mode || "block";

    let alias = (body.alias || "").trim() || (await nextAlias());
    const secret = (body.secret || "").trim();

    let error = null;
    if (!ALIAS_RE.test(alias)) {
      error = "Alias must be 2-32 characters: letters, numbers, dashes or underscores.";
    } else if (await ApiKey.findOne({ where: { alias } })) {
      error = `The alias '${alias}' is already taken. Aliases are global, try another.`;
    } else if (!secret) {
      error = "Paste the provider API key.";
    } else if (provider === "custom" && !baseUrl.trim()) {
      error = "A custom provider needs a base URL.";
    }

    if (error) {
      return res.status(400).render("key_new.html", {
        user: req.user,
        providers: PROVIDERS,
        suggested_alias: alias,
        created: null,
        error,
      });
    }

    const key = await ApiKey.create({
      userId: req.user.id,
      alias,
      label: label.trim(),
      provider,
      baseUrl: baseUrl.trim(),
      secretCiphertext: encryptSecret(secret),
      secretLast4: secret.slice(-4),
      passphraseHash: passphrase.trim() ? await hashSecret(passphrase.trim()) : null,
      active: true,
    });

    await KeyLimit.create({
      apiKeyId: key.id,
      rateLimit: Math.max(0, rateLimit),
      rateWindowSeconds: Math.max(1, rateWindowSeconds),
      rateMode: mode,
      spendCapUsd: Math.max(0.0, spendCapUsd),
      spendMode: mode,
    });

    res.render("key_new.html", {
      user: req.user,
      providers: PROVIDERS,
      suggested_alias: await nextAlias(),
      created: key,
      has_passphrase: Boolean(passphrase.trim()),
      error: null,
    });
  })
);

// ---------------- test connection ----------------

router.post(
  "/api/keys/:alias/test",
  requireAuth,
  asyncHandler(async (req, res) => {
    const key = await ownedKey(req.user, req.params.alias);
    if (!key.active) {
      res.set("Content-Type", "text/html");
      return res.send(
        '<span class="badge badge-bad">Revoked &mdash; re-activate this alias first</span>'
      );
    }

    const base = resolveBaseUrl(key.provider, key.baseUrl);
    const url = `${base}${probePath(key.provider)}`;
    const headers = authHeaders(key.provider, decryptSecret(key.secretCiphertext));

    const started = Date.now();
    let response;
    try {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), 15000);
      try {
        response = await fetch(url, { method: "GET", headers, signal: controller.signal });
      } finally {
        clearTimeout(timeout);
      }
    } catch (err) {
      res.set("Content-Type", "text/html");
      return res.send(
        `<span class="badge badge-bad">Could not reach provider &mdash; ${err.constructor.name}</span>`
      );
    }

    const ms = Date.now() - started;
    const responseText = await response.text();

    await KeyUsage.create({
      apiKeyId: key.id,
      path: probePath(key.provider),
      model: "",
      statusCode: response.status,
      costUsd: 0.0,
      latencyMs: ms,
      isTest: true,
      clientIp: "dashboard",
    });

    res.set("Content-Type", "text/html");
    if (response.status < 400) {
      return res.send(
        `<span class="badge badge-good">Success &mdash; provider responded ` +
          `${response.status} in ${ms}ms</span>`
      );
    }

    const detail = responseText.slice(0, 180).replace(/</g, "&lt;");
    return res.send(
      `<span class="badge badge-bad">Provider returned ${response.status}</span>` +
        `<pre class="probe-error">${detail}</pre>`
    );
  })
);

// ---------------- detail / settings ----------------

router.get(
  "/keys/:alias",
  requireAuth,
  asyncHandler(async (req, res) => {
    const key = await ownedKey(req.user, req.params.alias);
    const view = await keyView(key);

    const recent = await KeyUsage.findAll({
      where: { apiKeyId: key.id },
      order: [["createdAt", "DESC"]],
      limit: 50,
    });

    const keyAlerts = await Alert.findAll({
      where: { apiKeyId: key.id },
      order: [["createdAt", "DESC"]],
      limit: 20,
    });

    res.render("key_detail.html", {
      user: req.user,
      view,
      recent,
      alerts: keyAlerts,
      providers: PROVIDERS,
    });
  })
);

router.post(
  "/keys/:alias/limits",
  requireAuth,
  asyncHandler(async (req, res) => {
    const key = await ownedKey(req.user, req.params.alias);
    const body = req.body;

    let limit = await KeyLimit.findOne({ where: { apiKeyId: key.id } });
    if (!limit) {
      limit = await KeyLimit.create({ apiKeyId: key.id });
    }

    limit.rateLimit = Math.max(0, parseInt(body.rate_limit, 10) || 0);
    limit.rateWindowSeconds = Math.max(1, parseInt(body.rate_window_seconds, 10) || 60);
    limit.rateMode = body.rate_mode || "block";
    limit.spendCapUsd = Math.max(0.0, parseFloat(body.spend_cap_usd) || 0.0);
    limit.spendMode = body.spend_mode || "block";
    await limit.save();

    res.redirect(303, `/keys/${key.alias}`);
  })
);

router.post(
  "/keys/:alias/passphrase",
  requireAuth,
  asyncHandler(async (req, res) => {
    const key = await ownedKey(req.user, req.params.alias);
    const passphrase = (req.body.passphrase || "").trim();
    key.passphraseHash = passphrase ? await hashSecret(passphrase) : null;
    await key.save();
    res.redirect(303, `/keys/${key.alias}`);
  })
);

router.post(
  "/keys/:alias/rename",
  requireAuth,
  asyncHandler(async (req, res) => {
    const key = await ownedKey(req.user, req.params.alias);
    const newAlias = (req.body.new_alias || "").trim();
    const label = req.body.label || "";

    if (ALIAS_RE.test(newAlias)) {
      const clash = await ApiKey.findOne({
        where: { alias: newAlias, id: { [Op.ne]: key.id } },
      });
      if (!clash) {
        key.alias = newAlias;
      }
    }
    key.label = label.trim();
    await key.save();

    res.redirect(303, `/keys/${key.alias}`);
  })
);

// ---------------- soft delete ----------------

router.post(
  "/keys/:alias/revoke",
  requireAuth,
  asyncHandler(async (req, res) => {
    // Soft delete only: flip the flag, keep every usage row and alert.
    const key = await ownedKey(req.user, req.params.alias);
    key.active = false;
    key.revokedAt = new Date();
    await key.save();
    res.redirect(303, "/");
  })
);

router.post(
  "/keys/:alias/activate",
  requireAuth,
  asyncHandler(async (req, res) => {
    const key = await ownedKey(req.user, req.params.alias);
    key.active = true;
    key.revokedAt = null;
    await key.save();
    res.redirect(303, `/keys/${key.alias}`);
  })
);

router.post(
  "/keys/:alias/rotate",
  requireAuth,
  asyncHandler(async (req, res) => {
    const key = await ownedKey(req.user, req.params.alias);
    const secret = (req.body.secret || "").trim();
    if (secret) {
      key.secretCiphertext = encryptSecret(secret);
      key.secretLast4 = secret.slice(-4);
      await key.save();
    }
    res.redirect(303, `/keys/${key.alias}`);
  })
);

router.get(
  "/keys/:alias/curl",
  requireAuth,
  asyncHandler(async (req, res) => {
    const key = await ownedKey(req.user, req.params.alias);
    const base = settings.publicBaseUrl.replace(/\/+$/, "");
    const token = key.passphraseHash ? `${key.alias}<passphrase>` : key.alias;
    res.set("Content-Type", "text/html");
    res.send(
      `curl ${base}/proxy/${key.alias}/v1/chat/completions ` + `-H "Authorization: Bearer ${token}"`
    );
  })
);

module.exports = router;
